lib/ExtraCorres.thy

```diff
@@ -369,4 +369,53 @@ lemma corres_whileLoop_abs:
   apply (fastforce intro: wf_subset[OF whileLoop_terminates_wf[where C=C']])
   done
 
+
+text \<open>Some corres_underlying rules for monadic combinators\<close>
+
+lemma ifM_corres:
+  assumes   test: "corres_underlying srel nf nf' (=) A A' test test'"
+  and          l: "corres_underlying srel nf nf' rrel Q Q' a a'"
+  and          r: "corres_underlying srel nf nf' rrel R R' b b'"
+  and  abs_valid: "\<lbrace>A\<rbrace> test \<lbrace>\<lambda>c s. (c \<longrightarrow> Q s) \<and> (\<not> c \<longrightarrow> R s)\<rbrace>"
+  and conc_valid: "\<lbrace>A'\<rbrace> test' \<lbrace>\<lambda>c s. (c \<longrightarrow> Q' s) \<and> (\<not> c \<longrightarrow> R' s)\<rbrace>"
+  shows "corres_underlying srel nf nf' rrel A A' (ifM test a b) (ifM test' a' b')"
+  unfolding ifM_def
+  apply (rule corres_guard_imp)
+    apply (rule corres_split_forwards[OF test abs_valid conc_valid])
+    apply clarsimp
+    apply (intro conjI impI)
+     apply (rule l)
+    apply (rule r)
+   apply fastforce
```

Comment:
Could this be a `(fastforce intro: l r)`? (or `intro!`)

Reply:
After the `corres_split_forwards` step, the rest of the subgoals can be solved with `fastforce intro: assms`. Is that best?

Reply:
Yes, that looks nicer.